File: matlab/src/opt/+opt/OptimizationResult.m
classdef OptimizationResult < handle
    % OptimizationResult  Result from a single optimization run. Mirrors
    % native-Python line_solver.opt.results.OptimizationResult.

    properties
        objectiveValue = inf;
        variableValues            % Map name -> value (scalar or vector)
        constraintViolations      % Map name -> violation
        feasible = false;
        iterations = 0;
        solveTime = 0.0;
        modelEvaluations = 0;
        convergenceHistory = [];
        terminatedBy = '';
    end

    methods
        function obj = OptimizationResult()
            obj.variableValues = configureDictionary('string', 'cell');
            obj.constraintViolations = configureDictionary('string', 'double');
        end

        function tf = isFeasible(obj)
            tf = obj.feasible;
        end
        function v = getObjectiveValue(obj)
            v = obj.objectiveValue;
        end
        function v = getVariableValue(obj, name)
            if isKey(obj.variableValues, name), v = obj.variableValues{name}; else, v = []; end
        end
        function v = getConstraintViolation(obj, name)
            if isKey(obj.constraintViolations, name), v = obj.constraintViolations(name); else, v = 0.0; end
        end
        function v = getTotalViolation(obj)
            if numEntries(obj.constraintViolations) == 0, v = 0.0;
            else, v = sum(values(obj.constraintViolations)); end
        end
    end
end

File: jar/src/main/java/jline/opt/results/OptimizationResult.java
package jline.opt.results;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Result from a single optimization run. Mirrors native-Python
 * {@code line_solver.opt.results.OptimizationResult}: the optimal variable
 * values (by name), objective value, constraint status, and solver statistics.
 * Variable values are stored as {@code Object} because a variable decodes to a
 * scalar (server count, rate, population, priority) or a vector (routing).
 */
public class OptimizationResult {

    public double objectiveValue = Double.POSITIVE_INFINITY;
    public final Map<String, Object> variableValues = new LinkedHashMap<String, Object>();
    public final Map<String, Double> constraintViolations = new LinkedHashMap<String, Double>();
    public boolean feasible = false;
    public int iterations = 0;
    public double solveTime = 0.0;
    public int modelEvaluations = 0;
    public final List<Double> convergenceHistory = new ArrayList<Double>();
    public String terminatedBy = "";

    public boolean isFeasible() {
        return feasible;
    }

    public double getObjectiveValue() {
        return objectiveValue;
    }

    public Object getVariableValue(String name) {
        return variableValues.get(name);
    }

    public double getConstraintViolation(String name) {
        Double v = constraintViolations.get(name);
        return v == null ? 0.0 : v;
    }

    public double getTotalViolation() {
        double sum = 0.0;
        for (double v : constraintViolations.values()) {
            sum += v;
        }
        return sum;
    }

    public String toString() {
        String status = feasible ? "feasible" : "infeasible";
        return String.format("OptimizationResult(obj=%.4f, %s, iters=%d, evals=%d, time=%.2fs)",
                objectiveValue, status, iterations, modelEvaluations, solveTime);
    }
}
